### ISSUE TITLE
Add hydra support

### ISSUE THREAD

**Issue by RobinU434** [RobinU434/PyArgWriter#19]
Here is a simple rundown what ChatGPT had to say about it:

Combining `argparse` and `Hydra` is a useful approach when you want to manage configurations using Hydra while still maintaining some flexibility to override or add configurations through command-line arguments with `argparse`.

Below is a structured way to integrate `argparse` with `Hydra`:

#### 1. **Install Hydra**
Make sure you have `hydra-core` installed. If not, you can install it with:

```bash
pip install hydra-core
```

#### 2. **Define a Configuration File for Hydra**

Create a configuration file, say `config.yaml`, that Hydra will use. This file will define your default settings. For example:

```yaml
# config.yaml
database:
  user: admin
  password: secret
  host: localhost
  port: 5432

training:
  batch_size: 32
  epochs: 10
  learning_rate: 0.001
```

#### 3. **Setup the Python Script with `argparse` and `Hydra`**

Below is a sample script that shows how to combine `argparse` with `Hydra`:

```python
import argparse
from omegaconf import OmegaConf
import hydra
from hydra.core.config_store import ConfigStore

# Define the Hydra configuration structure
class Config:
    database: dict
    training: dict

cs = ConfigStore.instance()
cs.store(name="config", node=Config)

@hydra.main(version_base=None, config_path=".", config_name="config")
def main(cfg):
    # Initialize argparse
    parser = argparse.ArgumentParser(description="Combine argparse and Hydra example")

    # Define argparse arguments
    parser.add_argument("--learning_rate", type=float, help="Override learning rate")
    parser.add_argument("--epochs", type=int, help="Override number of epochs")
    parser.add_argument("--user", type=str, help="Override database user")

    # Parse known arguments (ignore arguments that Hydra has already handled)
    args, _ = parser.parse_known_args()

    # Override configuration with argparse values if provided
    if args.learning_rate:
        cfg.training.learning_rate = args.learning_rate
    if args.epochs:
        cfg.training.epochs = args.epochs
    if args.user:
        cfg.database.user = args.user

    # Print the final configuration
    print(OmegaConf.to_yaml(cfg))

    # Proceed with the rest of your program using the `cfg`
    # For example: access cfg.training.learning_rate or cfg.database.user
    print(f"Training for {cfg.training.epochs} epochs with learning rate {cfg.training.learning_rate}")
    print(f"Database user is {cfg.database.user}")

if __name__ == "__main__":
    main()
```

#### 4. **Explanation of Key Components**
- **Hydra Configuration (`config.yaml`)**: This file contains your default configuration. Hydra automatically loads this when you run the script.
- **`@hydra.main` Decorator**: This is a decorator from Hydra that initializes the configuration. The `config_path` should point to where your `config.yaml` is located.
- **`argparse` Integration**: The `argparse` arguments are parsed after Hydra loads the configuration, and you can override specific parameters from the command line.
- **Updating Configurations**: Use the parsed arguments from `argparse` to override values in the Hydra configuration. 

#### 5. **Running the Script**
You can run the script and override the configuration using both `Hydra` and `argparse`. For example:

```bash
# Using Hydra to override the config
python script.py training.learning_rate=0.01 database.user=new_user

# Using argparse to override specific values
python script.py --learning_rate 0.02 --user another_user
```

In both cases, the `Hydra` configuration will be merged with the `argparse` overrides.

#### 6. **Handling Conflicts**
If you need to prioritize one system over the other, be sure to establish the order of precedence when applying updates. In the example above, `argparse` overrides Hydra's configuration if both are provided.

### Advantages of Combining `argparse` and `Hydra`
- **Flexible Configuration**: `Hydra` manages complex configurations with YAML, and `argparse` allows quick command-line overrides.
- **Compatibility**: Easy to migrate older scripts using `argparse` to Hydra.
- **Clarity**: You get a clear YAML-based configuration file and still maintain the convenience of quick command-line tweaks.

### Tips
- Use `OmegaConf.to_yaml(cfg)` to visualize the final configuration.
- You can use `argparse.ArgumentParser(add_help=False)` if you want Hydra to handle the help messages.

This setup gives you the power of both `Hydra`'s configuration management and `argparse`'s straightforward CLI overrides.

When integrating `argparse` and `Hydra`, you can also take advantage of Hydra's **multirun** feature, which allows you to execute multiple configurations at once. This feature is particularly powerful for running hyperparameter sweeps or batch experiments. Below, I'll walk you through how to use `Hydra`'s multirun alongside `argparse`.

### What is Hydra's Multirun?
Hydra's `multirun` feature is used to run the same script multiple times with different configurations. Instead of using the regular `python script.py` command, you use `python script.py -m` to initiate a multirun. In multirun mode, Hydra will generate combinations of configuration values based on your specifications and run your script for each combination.

#### 1. **Example Configuration File for Multirun**

Suppose you have the following configuration file `config.yaml`:

```yaml
# config.yaml
training:
  batch_size: 32
  epochs: 10
  learning_rate: 0.001

model:
  type: "resnet"
  dropout: 0.3
```

#### 2. **Modify the Script to Use Multirun with `argparse`**

Here's a modified version of the previous example that supports multirun:

```python
import argparse
from omegaconf import OmegaConf
import hydra
from hydra.core.config_store import ConfigStore

# Define the Hydra configuration structure
class Config:
    training: dict
    model: dict

cs = ConfigStore.instance()
cs.store(name="config", node=Config)

@hydra.main(version_base=None, config_path=".", config_name="config")
def main(cfg):
    # Initialize argparse
    parser = argparse.ArgumentParser(description="Combine argparse and Hydra example with Multirun")

    # Define argparse arguments
    parser.add_argument("--learning_rate", type=float, help="Override learning rate")
    parser.add_argument("--epochs", type=int, help="Override number of epochs")
    parser.add_argument("--model_type", type=str, help="Override model type")

    # Parse known arguments (ignore arguments that Hydra has already handled)
    args, _ = parser.parse_known_args()

    # Override configuration with argparse values if provided
    if args.learning_rate:
        cfg.training.learning_rate = args.learning_rate
    if args.epochs:
        cfg.training.epochs = args.epochs
    if args.model_type:
        cfg.model.type = args.model_type

    # Print the final configuration for this run
    print(OmegaConf.to_yaml(cfg))

    # Proceed with the rest of your program using the `cfg`
    # For example: access cfg.training.learning_rate or cfg.model.type
    print(f"Training for {cfg.training.epochs} epochs with learning rate {cfg.training.learning_rate}")
    print(f"Model type is {cfg.model.type}")

if __name__ == "__main__":
    main()
```

#### 3. **Running Multirun Commands**

Now that the script is set up, you can use Hydra’s multirun features. Here's how to execute some typical scenarios:

##### **Basic Multirun Usage**
```bash
# Run the script with multiple learning rates
python script.py -m training.learning_rate=0.001,0.01,0.1
```

This command will run the script three times with `training.learning_rate` set to `0.001`, `0.01`, and `0.1`.

##### **Using Ranges in Multirun**
```bash
# Use a range for epochs
python script.py -m training.epochs=5,10,15
```

This will run the script three times with `training.epochs` set to `5`, `10`, and `15`.

##### **Combining Multiple Parameters**
```bash
# Sweep over multiple parameters
python script.py -m training.learning_rate=0.001,0.01 training.epochs=10,20
```

This will create four runs with all combinations:
1. `learning_rate=0.001` and `epochs=10`
2. `learning_rate=0.001` and `epochs=20`
3. `learning_rate=0.01` and `epochs=10`
4. `learning_rate=0.01` and `epochs=20`

#### 4. **Combining Multirun and `argparse` Overrides**

You can also override values using `argparse` while running a multirun. For example:

```bash
# Use argparse to set model type and Hydra for multirun on training parameters
python script.py -m --model_type cnn training.learning_rate=0.01,0.05 training.epochs=10,20
```

In this case:
- `model_type` is set to `cnn` through `argparse`.
- The multirun is performed over the combinations of `learning_rate` and `epochs`.

#### 5. **Handling Multirun Output**
When you perform a multirun, Hydra creates a separate output directory for each run. By default, the output directories are named something like:

```
outputs/2023-11-20/12-34-56/0
outputs/2023-11-20/12-34-56/1
outputs/2023-11-20/12-34-56/2
...
```

Each directory will have a copy of the configuration used for that specific run (`.hydra` folder) and any output your script generates.

### Additional Tips

- Use **Hydra Overrides** with `argparse` to handle different configurations directly from the command line, making it more flexible for experiments.
- Use the `hydra.sweep.dir` and `hydra.sweep.subdir` options to control the output directory structure for multirun.
- To manage combinations efficiently, you can use **YAML Anchors** in your configuration or define lists of values for a parameter directly in `config.yaml`.

### Example: Hydra Multirun with Parallel Jobs
You can also run multiple jobs in parallel by specifying the `-p` option:

```bash
# Run 4 jobs in parallel
python script.py -m training.learning_rate=0.001,0.01,0.1 training.epochs=10,20 -p hydra/launcher=basic hydra.launcher.n_jobs=4
```

This command runs the multirun with 4 jobs in parallel, making it faster when you have many combinations to test.

By using Hydra’s multirun feature, you can efficiently explore hyperparameters and configuration setups while still having the flexibility to manually override specific values via `argparse`. This makes it a powerful tool for experimentation and tuning.

**Comment by RobinU434** [comment]
Add this feature as a flag for the pyargwriter: 

```bash
pyargwriter generate-argparser --input xxxx --output xxxx --hydra
```

**Comment by RobinU434** [comment]
Check for arguments already named from hydra
